File: 03.CPP/0424.cpp
#include <iostream>
#include <string>
#include <vector>

class Solution {
public:
    int characterReplacement(std::string s, int k) {
        int n = s.size();

        int left = 0;
        int ans = 0;
        int maxFreq = 0;

        std::vector<int> freq(26, 0);

        for (int right = 0; right < n; right++) {
            int idx = s[right] - 'A';
            freq[idx]++;
            maxFreq = std::max(maxFreq, freq[idx]);

            while (((right - left + 1) - maxFreq) > k) {
                int idxL = s[left] - 'A';
                freq[idxL]--;
                left++;
            }

            ans = std::max(ans, right - left + 1);
        }

        return ans;
    }
};

int main() {
    Solution sol;

    std::string s = "ABAB";
    int k = 2;

    int result = sol.characterReplacement(s, k);
    std::cout << result << "\n";

    return 0;
}

File: 01.Python/0424.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        n = len(s)

        left = 0
        maxFreq = 0
        ans = 0

        freq = [0] * 26

        for right in range(n):
            idx = ord(s[right]) - ord('A')
            freq[idx] += 1
            maxFreq = max(maxFreq, freq[idx])

            diff = (right - left + 1) - maxFreq
            while diff > k:
                idxL = ord(s[left]) - ord('A')
                freq[idxL] -= 1
                left += 1
            
            ans = max(ans, (right - left + 1))

        return ans

sol = Solution()

s = "ABAB"
k = 2

result = sol.characterReplacement(s, k)
print(result)